src/ripple/overlay/impl/ProtocolVersion.cpp

```diff
@@ -0,0 +1,184 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/overlay/impl/ProtocolVersion.h>
+#include <ripple/beast/core/LexicalCast.h>
+#include <ripple/beast/rfc2616.h>
+#include <boost/function_output_iterator.hpp>
+#include <boost/regex.hpp>
+#include <algorithm>
+#include <functional>
+
+namespace ripple {
+
+/** The list of protocol versions we speak and we prefer to use.
+
+    @note The list must be sorted in strictly ascending order (and so
+          it may not contain any duplicates!)
+*/
+constexpr
+ProtocolVersion const supportedProtocolList[]
+{
+    { 1, 2 },
+    { 2, 0 }
+};
+
+// This ugly construct ensures that supportedProtocolList is sorted in strictly
+// ascending order and doesn't contain any duplicates.
+// FIXME: With C++20 we can use std::is_sorted with an appropriate comparator
+static_assert(
+    []() constexpr -> bool
+    {
+        auto const len = std::distance(
+            std::begin(supportedProtocolList),
+            std::end(supportedProtocolList));
+
+        // There should be at least one protocol we're willing to speak.
+        if (len == 0)
+            return false;
+
+        // A list with only one entry is, by definition, sorted so we don't
+        // need to check it.
+        if (len != 1)
+        {
+            for (auto i = 0; i != len - 1; ++i)
+            {
+                if (supportedProtocolList[i] >= supportedProtocolList[i+1])
+                    return false;
+            }
+        }
+
+        return true;
+    }(), "The list of supported protocols isn't properly sorted.");
+
+
+std::string
+to_string(ProtocolVersion const& p)
+{
+    // The legacy protocol uses a different name. This can be removed when we
+    // migrate away from it and require 2.0 or later.
+    if (p == ProtocolVersion{ 1, 2 })
```

Comment:
It's only used three times, but you could define a `constexpr ProtocolVersion legacyVersion{ 1, 2 }` and use it in those few places. When we migrate away from it, remove the variable.